src/neuroglancer/segmentation_graph/source.ts

```diff
@@ -23,18 +23,18 @@ import {RenderLayerTransformOrError} from 'neuroglancer/render_coordinate_transf
 import { ChunkManager } from 'neuroglancer/chunk_manager/frontend';
 import { ChunkedGraphLayerDisplayState } from 'neuroglancer/sliceview/chunked_graph/frontend';
 
-export enum VISIBLE_SEGMENT_TYPE {
-  SIMPLE_EQUIVALENCES = 1,
-  HIGH_BIT_REPRESENTATIVE_EXCLUDED,
-  HIGH_BIT_REPRESENTATIVE_ONLY
+export enum VisibleSegmentType {
+  SIMPLE_EQUIVALENCES = 1,          // all members of a set are accessed as individual elements
+  HIGH_BIT_REPRESENTATIVE_EXCLUDED, // equivalences with the highest bit set are ignored
+  HIGH_BIT_REPRESENTATIVE_ONLY      // only the representive
```

Comment:
HIGH_BIT_REPRESENTATIVE_ONLY doesn't seem to be the correct way to describe the option we need for graphene. VISIBLE_SEGMENT_ONLY would be more fitting. We only want to interact with the visibleSegment set. Graphene replaces non root segments in that set with root segments.

Graphene has it's own isBaseSegmentId. It would probably makes sense to move that into a method on SegmentationGraphSourceConnection that graphene can override.

Another possible name could work BASE_SEGMENT_EXCLUDED/BASE_SEGMENT_ONLY

Reply:
Currently this enum controls two things:

- The behavior of DisjointUint64Sets, namely which element of the set is considered the representative (either min or max)
- Which segments in the visible equivalence sets are iterated over by `forEachVisibleSegment` in `segmentation_display_state/base.ts`.  That is used for the existing mesh, skeleton, and annotation layers.

Maybe it wold be better to decouple them, so that we just have a MIN_REPRESENTATIVE / MAX_REPRESENTATIVE option for DisjointUint64Sets, and something else for `forEachVisibleSegment`.

I believe for graphene you want `forEachVisibleSegment` to only iterate over the representative.  So we could potentially have:

VisibleSegmentEquivalencePolicy with

INCLUDE_ALL
EXCLUDE_REPRESENTATIVE
REPRESENTATIVE_ONLY

Would that work?

Reply:
Yes, those names make sense to me. Though it may make sense to additionally modify forEachVisibleSegment to ignore base segments in the graphene REPRESENTATIVE_ONLY case. Currently there is an extraneous mesh manifest request made for the original segment id before graphene swaps it with the root id. Since it is a non root id, the manifest just returns an empty array. It is not significant.

VisibleSegmentType is also used in the MergeSegmentsTool. We may support MergeSegmentsTool in the future though our first iteration of graphene proofreading will probably only use our graphene specific merge tool.

Reply:
That is a good point.  One tricky thing if you need a call to the graph source to determine which segments to include is that the graph source currently exists only on the frontend, but this function is used on the backend as well.

Reply:
Would you be opposed to having a back-end counterpart to the chunked graph?

One issue I find with our implementation is that our chunked_graph had graphene logic abstracted out. I think the reason was because it was based on sliceview so it was moved with the rest of the sliceview code. I would be surprised if anyone else decided to use chunked_graph and I don't think it is generic enough to support other implementations.

I am currently assuming it is a graphene datasource in chunked_graph backend because of the isBaseSegment call. Do you think I should moved chunked_graph into graphene? and maybe combine GrapheneChunkedGraphChunkSource and ChunkedGraphChunkSource.

Reply:
Yes I think combining could make sense, as could having a backend counterpart for a graph connection.